## Shiny UI for MB nanoString Classification
## Machine learning model and classifier code: Dr Reza Rafiee
## Adaptation and Shiny code by Dr Reza Rafiee, Dr Matthew Bashton 2014-2017

#install.packages("shiny")
library(shiny) # load shiny at beginning at both scripts
#install.packages("shinythemes")
library(shinythemes)
shinyUI(fluidPage(theme = shinytheme("united"),
                  
                  titlePanel("NIMBLE: NanostrIng MedulloBlastoma cLassifiEr"), # give the interface a title
                  
                  br(),
                  sidebarLayout(
                    
                    sidebarPanel( # all the UI controls go in here
                      width=3,
                      fileInput('file1', 'NanoString CSV file upload:', #'MassARRAY CSV file upload:',
                                accept=c('text/csv',
                                         'text/comma-separated-values,text/plain',
                                         '.csv')
                      ),
                      helpText("NIMBLE will classify NanoString medulloblastoma gene expression data in to one of four molecular subgroups"),
                      "Download test data to try classifier:", a(href="test_samples.csv", "Test CSV file", target="_blank")
                      
                    ), # End sidebarPanel
                    
                    
                    mainPanel( # all of the output elements go here
                      tabsetPanel(
                        id = 'sequenom',
                        tabPanel('Classification Table', dataTableOutput('classification_table'), br(), helpText("Unclassifiable samples are those for which a confident subgroup call could not be made"), br(), downloadButton('downloadClassification', 'Download table as .csv'),p(), br(), textOutput("time")),
                        tabPanel('Classification Plot', plotOutput("classifierPlot", height = "640", width = "auto"), br(), helpText("Unclassifiable and Normalization QC failed samples are not shown in this plot.  Boxes show the confidence interval for subgroup assignment generated by bootstrapping, and the individual data points represent the final probability associated with each subgroup call."), br(), downloadButton('PlotDownload', 'Download plot as high-res .png'), p()),
                        #tabPanel('Informative Probes Table', dataTableOutput('mp'), br(), downloadButton('downloadMissing', 'Download table as .csv')),
                        tabPanel('Sample QC', br(), p(), textOutput("fs"), br(), textOutput("fc")),
                        #tabPanel('Bisulphite Conversion Efficiency', dataTableOutput("BS_Eff"), br(), downloadButton('downloadBS_Eff', 'Download table as .csv')),
                        tabPanel('Gene Expression', dataTableOutput('Beta'), br(), downloadButton('downloadBeta', 'Download table as .csv')),
                        tabPanel('About',
                                 h4("NIMBLE version 3.2.1-1"),
                                 br(),
                                 p("Machine learning model and classifier code: ", a("Reza Rafiee", href="mailto:Gholamreza.Rafiee@newcastle.ac.uk?subject=Sequenom classifier website")),
                                 p("Shiny web app code and adaptation: ", a("Reza Rafiee", href="mailto:Gholamreza.Rafiee@newcastle.ac.uk?subject=nanoString classifier website"),", Matthew Bashton"),
                                 br(),
                                 h4("Overview"),
                                 p("NIMBLE will classify nanoString medulloblastoma gene expression data into one of four molecular subgroups: WNT, SHH, Group 3 and Group 4."),
                                 p("In summary the classifier works as described below:"),
                                 tags$ol(
                                  tags$li("We use 19 different gene expression read count (raw data) in a", a("NanoString assay", href = "http://www.nanostring.com/")), p(),
                                  tags$li("NanoStringNorm R package is used to normalize the input raw data"), p(), 
                                  tags$li("A multi-class optimised ", a("Support Vector Machine", href = "https://en.wikipedia.org/wiki/Support_vector_machine"), "(SVM) validated and trained on our extensive RNA-seq medulloblastoma cohort is used to robustly assign a subgroup to samples by their 19 gene expression values."), p(),
                                   tags$ul(
                                     tags$li("Our SVM is validated using a bootstrapping technique via 1,000 random iterations of 90% of the training set, confidence interval derived from this is plotted on the Classification Graph as a box plot."), p(),
                                     tags$li("The final probability assignment for a subgroup call is made by creating an SVM model with the whole RNA-seq training set; these probabilities are given in the Classification Table in the initial tab."), p(),
                                     tags$li("Calls made with a probability below our predefined threshold are considered unreliable and samples will be labeled as Unclassifiable in the Classification Table, these samples will not be plotted in the Classification Graph."), p()
                                   ),
                                   tags$li("Various post processing and formatting operations on the data take place with the interactive website being implemented in the R", a("Shiny", href = "http://shiny.rstudio.com"), "reactive web application framework."), p()
                                 ), # End ol
                                 p("For a typical dataset with 14 samples this whole computational procedure will take around 4 seconds - total classification time is given below the Classification Graph. "),
                                 p("For more detailed explanation of our classifier including various optimisation and validation exercise see our manuscript and corresponding supplementary information (manuscript in preparation)."),
                                 p(), br(),
                                 h4("Reference"),
                                 p("A manuscript is in preparation."),
                                 p(), br(),
                                 h4("Download"),
                                 "The R code for this", a("Shiny",target="_blank", href = "http://shiny.rstudio.com/"), "based website including training and validation cohorts can be downloaded from", a("GitHub", href = "https://github.com/RRafiee/NIMBLE/"), "the website can also be run locally using", a("Rstudio", href = "https://www.rstudio.com/"), "instructions and dependancies are outlined on GitHub.",
                                 p(), br(),
                                 h4("Funding"),
                                 p("NIMBLE development was funded by a Cancer Research UK program grant.")
                        ),

                        tabPanel('Help',h4("How to use our Classifier"),
                                 p("NIMBLE will classify nanoString medulloblastoma gene expression data into one of four molecular subgroups.  To use the classifier follow the steps outlined below:"),
                                 tags$ol(
                                   tags$li("A Comma separated value (.csv) file produced by a nanoString assay is needed as input to use the classifier.  If you would like to test drive the classifier,  or would like to see how it should be formatted a test file can be downloaded using the link in the grey box on the left."),
                                   p(), img(src = "step1.png"), p(), br(),
                                   tags$li("A nanoString .csv file can then be uploaded by clicking on the 'Chose File' or 'Browse...' (browser dependent) button on the left, once uploaded the classification happens automatically."),
                                   p(), img(src = "step2.png"), p(), br(),
                                   tags$li("By default the Classification Table output is preselected and will present you with a four subgroup Medulloblastoma classification for each of your samples.  Other tabs presenting other information can then be accessed by clicking their names present at the top of the main panel."),
                                   p(), img(src = "step3.png"), p(), br(),
                                   tags$li("The contents of Tables can be downloaded by clicking the grey download button, these .csv files can then be loaded into Excel or other spreadsheet software if required."),
                                   p(), img(src = "step4.png"), p(), br(),
                                   tags$li("The Classification Plot can also be downloaded as a .png by clicking on the grey Download button at the bottom of the Classification Plot tab."),
                                   p(), img(src = "step5.png"), br()
                                 ), # End of list
                                 p(), br(),
                                 h4("Input file format"),
                                 "The input file used for this version of NIMBLE is exported from a nanoString nCounter software. We provide an example file in", a(href="test_samples.csv", "test_samples.csv", target="_blank"), p(),
                                 p("Gene names we use are:
                                   DKK2, EMX2, GAD1, TNC, WIF1 (WNT subgroup genes), 
                                   ATOH1, EYA1, HHIP, SFRP1 (SHH subgroup genes),
                                   GABRA5, IMPG2, MAB21L2, NPR3, NRL (Grp3 subgroup genes),
                                   EOMES, KCNA1, KHDRBS2, RBM24, UNC5D (Grp4 subgroup genes)."),
                                 p(), br(),
                               h4("Suppport"),
                               "If you have any issues with using NIMBLE please contact ",
                               a("Reza Rafiee.", href="mailto:Gholamreza.rafiee@newcastle.ac.uk?subject=nanoString classifier website")
                               ) # End of Help tab
                               
                    ), # End of tabsetPanel
                    br(),
                    hr(),
                    p("WARNING: NIMBLE is for research use only, and should only be used on samples with a confirmed histopathological diagnosis of medulloblastoma."),
                    hr(),
                    img(src = "nicr.png"), img(src = "ncl.png"),
                    br()
                  ) # End of mainPanel

  ) # End sidebarLayout

) # End fluidPage

) # End shinyUI